### PR TITLE
Docker blog post

### PR DESCRIPTION
**Pull request by choffmeister**
@khauser could you please take a look at this post and tell me, if you think it is from a technical point of view ok?

### CODE REVIEW

**Comment by khauser** [comment]
very fine!

**Comment by choffmeister** [comment]
OK, so @BirgitBader please proof-read this post.

**Comment by BirgitBader** [comment]
@choffmeister I'm done. Please include my feedback.

**Comment by BirgitBader** [comment]
I'm done re-reviewing it.

**Comment by choffmeister** [comment]
@steffimeier @maletzke Please review.

File: _posts/2015-06-11-docker.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+# The challenge
+
+A simplified version of our current project is the following: We are developing a [NodeJS][nodejs]-based application that uses a [Java][java]-based backend for storing data which itself uses a [MySQL][mysql] database. Pretty standard stuff. So far we deployed the backend service via continuous deployment to a central staging machine and this address was hard coded into our frontend project. It worked but still could be improved: Backend changes that were not backwards compatible necessarily blocked our frontend developers and forced them, to adapt to these changes before continuing other work.
+
+# The solution
+
+To improve this, we decided use Docker to make starting everything on our local machine easy.
```
Comment (maletzke) [inline _posts/2015-06-11-docker.md]: "we decided use Docker" --> "we decided to use Docker"

File: _posts/2015-06-11-docker.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    MYSQL_DATABASE: project
+{% endhighlight %}
+
+The file defines our three services, the frontend, the backend and the database. The Docker image to be used is defined for every service as well as the links, i.e. the frontend needs the backend and the backend needs the database. The frontend's container port `8080` gets exposed.
+
+That's it. A developer just needs this single YAML file and then has to run `docker-compose up`. Docker will take care of downloading the images if needed, starting the containers and linking them together. The developer gets to see the live standard output of all three services mixed into a single stream.
+
+## Conclusion
+
+It is quite a long way to get an existing system into Docker containers. But the way is well worth it. We did not just solve the problem that arised from having a central staging machine, but we also found several bugs, that were unnoticed due to the fact, that all the addresses were fixed before. Using Docker, the services receive new IP addresses on every (re)start. So we were pretty much forced to really think about what network services we expose and what network services we need to consume in order to make the system run.
```
Comment (maletzke) [inline _posts/2015-06-11-docker.md]: "we also found several bugs" --> i would prefer "we also found a few bugs" ;)

**Comment by choffmeister** [comment]
@RealMattMurdoch Please do the final proof-reading. Thanks.

File: _posts/2015-06-11-docker.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+In a nutshell, one can describe Docker as [Linux containers][lxc] with some extra convenience.
+Docker can be classified somewhere in between fully fledged virtual machines and a simple filesystem virtualisation with chroot.
+Since it is not yet clear, if Docker will find it's way into our production deployment, this post only concentrates on how our development workflow benefits from Docker.
+
+# The challenge
+
+A simplified version of our current project is the following: We are developing a [NodeJS][nodejs]-based application that uses a [Java][java]-based backend for storing data which itself uses a [MySQL][mysql] database.
+Pretty standard stuff.
+So far we deployed the backend service via continuous deployment to a central staging machine and this address was hard coded into our frontend project.
+It worked but still could be improved: Backend changes that were not backwards compatible necessarily blocked our frontend developers and forced them, to adapt to these changes before continuing other work.
```
Comment (RealMattMurdoch) [inline _posts/2015-06-11-docker.md]: "It worked, but could still be improved" sounds better

"necessarily" - you mean unnecessarily right? (unnötigerweise)

File: _posts/2015-06-11-docker.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+# The solution
+
+To improve this, we decided to use Docker to make starting everything on our local machine easy.
```
Comment (RealMattMurdoch) [inline _posts/2015-06-11-docker.md]: "easier"

File: _posts/2015-06-11-docker.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+RUN npm install
+RUN npm run build
+
+CMD ["node", "index.js"]
+{% endhighlight %}
+
+The four main steps here are:
+
+1. Start from the `node:0.12.3` base image from the [offical Docker Hub][docker-hub].
+2. Add the current directory (which is the project folder containing our frontend project sources) to the container.
```
Comment (RealMattMurdoch) [inline _posts/2015-06-11-docker.md]: I do not believe source (as in source code) has a plural, although I am not sure. I would say:
"(which is the project folder containing our frontend project source code)"

**Comment by choffmeister** [comment]
So, integrated all review annotation.

@BirgitBader @steffimeier @maletzke @RealMattMurdoch are there any more objections, or can this be merged and published?

**Comment by choffmeister** [comment]
There you go.

**Pull request merged**